### ISSUE TITLE
Dead code elimination should happen before codegen

### ISSUE THREAD

**Issue by paf31** [purescript/purescript#2970]
Extracted from #2963

Instead of doing DCE in `purs bundle`, we could do this instead:

- Provide a tool to extract dependencies from the functional core, and save the graph as a file
- Allow `purs build` to take that file as an input, and only generate code which is reachable from a set of entry points.

So the workflow would look like this:

- Build as usual using `purs build --dump-corefn`
- Make changes (using fast rebuilds and `--dump-corefn`)
- When ready to make a bundled build, run the new tool against the functional core, and rerun `purs build` using the output file, to produce a new set of commonjs modules with minimal definitions.
- Run the commonjs modules through psc-bundle, browserify, closure or whatever to turn it into a bundle.

I think this approach has lots of benefits:

- We can do dead code elimination in new backends
- The graph traversal code can be more reliable since it works on the more principled functional core and not JS
- We don't have to make guesses about which `module.export = ...` lines we can safely ignore.
- We can reuse work if we need to make multiple bundles.

The main downside of this approach is that we can't remove JS functions in foreign modules without parsing them, so that would have to be left to some other tool. Maybe Closure is capable of this?

**Comment by coot** [comment]
I think there is no need for a new format, we could reuse `CoreFn` for the output of the _DCE tool_:
* output the modified `CoreFn` output (in a possibly new directory, other than `output`, maybe `build` by default)
* add the `purs build` command that would take that output, read into `CoreFn` and generate `CommonJS` modules (the same way modules are generated now from `CoreFn`)

If one needs to build multiple bundles manually one can run this _DCE tool_ and output the DCE-ed `CoreFn` representation to different directories and use `purs build` for each of them. The over head here is that one might end-up with bundles that contain the same modules, but we're walking into build tool responsibility.

With _webpack_'s _CommonChunksPlugin_ one will have to run the _DCE tool_ & `pure build` just once, and then let _weback_ and _CommonChunksPlugin_ to decide how to glue `CommonJS` modules into chunks.

Later we could add our own common chunks splitting: that would also use only a single pass of _DCE tool_ and then run `purs build` with a set of entry points and a flag to extract common chunks into separate module.

 I'd be happy to start with it, and have a reasonable road path :)

**Comment by paf31** [comment]
I think it would be good to try out that idea in a branch and see how it goes.

**Comment by coot** [comment]
I found [purescript-corefn](https://github.com/paulyoung/purescript-corefn). In this case this could be implement in purescript but I am equally happy to do that in Haskell.
I don't think performance is really important here, since this will run mostly only for production builds.  Though doing this in Haskell might be better for other backends: no dependency on `node`.

Types of foreign declaration are [lost](https://github.com/purescript/purescript/blob/master/src/Language/PureScript/CoreFn/ToJSON.hs#L47) in the json represantation of `CoreFn` so it cannot simply be parsed back into `Module ()` type.

Btw. what is the usage of the type parameter  in `Expr a`, `Module a`, etc.

**Comment by paf31** [comment]
I think it is the annotation type, for source positions and things like that.

**Comment by coot** [comment]
ok, this part is also lost in the json representation, so for now any transformation on `CoreFn` read from the json representation will lose source positions. But anyway that can be added later.

**Comment by coot** [comment]
I started working on this issue [here](https://github.com/coot/purescript/tree/dce-tool).  So far I have [parser](https://github.com/coot/purescript/blob/dce-tool/src/Language/PureScript/CoreFn/FromJSON.hs) of `CoreFn`'s json representation. :)

**Comment by natefaubion** [comment]
Should this be closed? @coot released `zephyr`, which does dce on corefn, and works really well (I use it in production).

**Comment by hdgarrood** [comment]
I think there are still some unresolved questions:

- do we want to move `zephyr` into the compiler? 
- do we want to deprecate and eventually remove the current `purs bundle` (possibly replacing it with a more `zephyr`-like approach)?

If we think the `zephyr` approach is better (and it clearly does have a lot of benefits) then I don't think it makes a whole lot of sense for us to continue to support `purs bundle` inside the compiler, and additionally it's not great UX that users have to seek out and install a separate tool. I feel like the question of "why are my purescript bundles so big compared to other compile-to-js languages / js frameworks" comes up fairly often in Slack, and `zephyr` usually ends up being suggested in those threads; it seems plausible that we can make the getting started experience quite a bit more pleasant for people who care about bundle size by having the most effective dead code elimination work out-of-the-box.

**Comment by thomashoneyman** [comment]
A 👍 for moving `zephyr` into the compiler, as I suspect any PureScript web application that wants to get its bundle size down will eventually start using it. For example, a blank Halogen project will bundle at somewhere around 684kb, but running Zephyr will reduce that by something like 400kb because of all the unused `purescript-web` library code.

Zephyr doesn't have great discoverability as a separate project and so I'm sure many users who _would_ have used it end up settling for larger bundle sizes.

On the other hand, that would require some confidence that no other DCE tool will come along, be more effective, but be impossible to use because the compiler already blessed a particular solution. Given that there's been no competition to Zephyr in this space in the two years since it came out it might be premature to optimize for this being super flexible.

**Comment by hdgarrood** [comment]
Is there anything that makes you think moving Zephyr into the compiler could make it harder to create a separate DCE tool? If we were to move Zephyr into the compiler, we (or at least I) wouldn't want it to run unconditionally on `purs compile`; we'd only run it if the user has specifically asked for dead code to be eliminated, so the current `purs compile` API should still work the same way and therefore it should still be possible to prototype other approaches.

**Comment by thomashoneyman** [comment]
No, you're right: moving Zephyr wouldn't affect making other DCE tooling, as the `purs compile` API would be unchanged.

**Comment by f-f** [comment]
In the issue mentioned just above this comment I had the thought to bundle zephyr into Spago, but it makes so much more sense to have it in the compiler instead. This is all to say that this gets a +1 from me :slightly_smiling_face: 

Is there anything blocking this? I.e. what's the criteria to move it from "idea" to "intend to implement"?

**Comment by chexxor** [comment]
I am interested in what the name of the command to invoke this DCE would be called. In the JavaScript community, it's often just part of a thing called "bundling", whereas in old languages/compilers, such as GCC, it is simply a part of "optimized" build.

On a related tangent is where the GCC calls a "statically linked" compilation, JavaScript land calls it "bundling".

I don't mind too much which terms are used, but it's an interesting change in terminology over the decades which the name-chooser of this might find helpful to consider when choosing names. ☺️

**Comment by hdgarrood** [comment]
It's not quite the same as an optimized build. Optimization usually refers to transformations which aim to make the code run faster, like inlining; the goal here is to remove code which isn't used in order to reduce the size of the bundle. I think it's a bit early to start bikeshedding about what it's called, in any case.

The main blocker is that it's not entirely clear what we should do about dead code elimination for FFI modules with this approach. As far as I can tell, the options are:
* keep the existing `purs bundle` code in the compiler. This way, we can do very good DCE on corefn, and decent DCE on FFI modules, but now we have two separate DCE implementations to maintain.
* get rid of the existing `purs bundle` code entirely, and declare that DCE for JS is no longer the compiler's responsibility,
* get rid of just the part of `purs bundle` that does DCE on JS, and keep the rest of it as a convenience (e.g. for generating bundles which you can easily call into via `PS['Prelude'].whatever`)

I'm really not sure which I prefer. One of the main benefits of ES modules is supposed to be that they are easier to do dead code elimination on, and DCE tools are something that the JS community has invested quite heavily in over the last few years, so arguably we shouldn't bother trying to retread the same ground (with the current `purs bundle`). On the other hand, it is arguably a good thing that people can install the compiler and not have to install anything else in order to have reasonably small bundles. The core libraries use a lot of FFI so an approach that doesn't handle FFI code can't really be considered acceptable for most use cases.

I think it's also possible that we will ship ES modules and find that existing JS tooling becomes competitive with zephyr once the ecosystem has adjusted and is using ES modules for FFI, in which case I'm not sure we would need to support dead code elimination in the compiler at all. Perhaps we should revisit this at that point.

**Comment by garyb** [comment]
I would also reiterate the point that all the JS bundlers that tree shake are extremely slow (last time I tried at least, ES modules may help with that) compared to `purs bundle`, which isn't all that fast itself. 

I suppose that's perhaps not too much of a concern though, because you'd only use it on production bundles.

**Comment by hesxenon** [comment]
> I would also reiterate the point that all the JS bundlers that tree shake are extremely slow (last time I tried at least, ES modules may help with that) compared to `purs bundle`, which isn't all that fast itself.

Just my two cents on that: ESbuild is pretty fast and does ok treeshaking. I'm using it in most of my projects now and would see that as a reference when it comes to speed. (read: please make it fast, pretty please?)

**Comment by rhendric** [comment]
#4255 has landed, the bundler is dead, and we're planning on telling people to use esbuild (and/or call it from spago). Are there any threads here that this outcome doesn't resolve?

**Comment by JordanMartinez** [comment]
> Are there any threads here that this outcome doesn't resolve?

It sounds like moving `zephyr` into the compiler might still be desirable, at least for the benefit of other backends that don't have tools like `esbuild`. Also, this somewhat relates to the ideas proposed in #3339.

**Comment by MaybeJustJames** [comment]
If #4092 can be finished off, there are 2 paths forward as mentioned in #3339:
1. `purs compile --codegen corefn` + `$optimizer` + `purs codegen`
2. `purs compile --optimizer ./myoptimizer`
3. `purs compile --optimize` which would run zephyr

Options 1,2 need not require zephyr in the compiler, but 3 would. Is there some agreement on the path forward?
In my mind option 1 is preferable due to less complexity in the compiler.
